noir-projects/aztec-nr/aztec/src/oracle/aes128_decrypt.nr

```diff
@@ -0,0 +1,122 @@
+/// Decrypts a ciphertext, using AES128.
+///
+/// Returns a padded plaintext, of the same size as the input ciphertext.
+/// Note that between 1-16 bytes at the end of the returned plaintext will be pkcs#7 padding.
+/// It's up to the calling function to identify and remove that padding.
+/// See the tests below for an example of how.
+/// It's up to the calling function to determine whether decryption succeeded or failed.
+/// See the tests below for an example of how.
+unconstrained fn aes128_decrypt_oracle_wrapper<let N: u32>(
+    ciphertext: [u8; N],
+    iv: [u8; 16],
+    sym_key: [u8; 16],
+) -> [u8; N] {
+    aes128_decrypt_oracle(ciphertext, iv, sym_key)
+}
+
+#[oracle(aes128Decrypt)]
+unconstrained fn aes128_decrypt_oracle<let N: u32>(
+    ciphertext: [u8; N],
+    iv: [u8; 16],
+    sym_key: [u8; 16],
+) -> [u8; N] {}
+
+mod test {
+    use crate::{
+        encrypted_logs::encrypt::aes128::derive_aes_symmetric_key_and_iv_from_ecdh_shared_secret_using_sha256,
+        utils::point::point_from_x_coord,
+    };
+    use super::aes128_decrypt_oracle_wrapper;
+    use std::aes128::aes128_encrypt;
+
+    #[test]
+    unconstrained fn aes_encrypt_then_decrypt() {
+        let ciphertext_shared_secret = point_from_x_coord(1);
+
+        let (sym_key, iv) = derive_aes_symmetric_key_and_iv_from_ecdh_shared_secret_using_sha256(
+            ciphertext_shared_secret,
+        );
+
+        let plaintext: [u8; 10] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
+
+        let ciphertext = aes128_encrypt(plaintext, iv, sym_key);
+
+        let received_plaintext = aes128_decrypt_oracle_wrapper(ciphertext, iv, sym_key);
+        let padding_length = received_plaintext[received_plaintext.len() - 1] as u32;
```

Comment:
I don't understand this example, why is the padding length in the received plaintext?

Reply:
PKCS#7 padding standard says to pad a plaintext to the next multiple of 16 bytes with repeated bytes of the length of the padding.
- If plaintext is 0 bytes: Pad with 0x10 (16) sixteen times.
- If plaintext is 1 byte: Pad with 0x0f (15) fifteen times.
- If plaintext is 2 bytes: Pad with 0x0e (14) fourteen times.
- ...
- If plaintext is 15 bytes: Pad with 0x01 one time.

So the final byte always tells you how long the padding is.

Reply:
I added this explanation as a comment in #12065